File: hcl.js
// This is not a general purpose HCL formatter
// We know we're only doing a single resource
// And we also know we can rely on `terraform fmt` to tidy up after us

exports.buildHcl = function buildHcl(resourceType, name, resource) {
  const lines = [];

  lines.push(`resource ${value(resourceType)} ${value(name)} {`);
  lines.push(objectProperties(resource));
  lines.push('}');

  return lines.join("\n");

  function objectProperties(obj) {
    Object.keys(obj).forEach(key => {
      const val = obj[key];
      const type = typeof val;
      if (['number', 'string', 'boolean'].includes(type)) {
        primitve(key, val);
      } else if (!val) {
        primitve(key, "");
      } else if (Array.isArray(val)) {
        val.forEach((item) => object(key, item));
      } else {
        object(key, val);
      }
    });
  }
  function primitve(key, val) {
    lines.push(`${key} = ${value(val)}`);
  }
  function object(key, val) {
    lines.push(`${key} {`);
    lines.push(objectProperties(val));
    lines.push(`}`);
  }
  function value(val) {
    return JSON.stringify(val);
  }
}

File: hcl.py
import json


def build(resource_type, resource_name, resource):
    lines = []

    lines.append("resource %s %s {" %
        (value(resource_type), value(resource_name))
    )
    lines.extend(object_props(resource))
    lines.append("}")

    return "\n".join(lines)


def value(val):
    return json.dumps(val)


def object_props(obj):
    lines = []

    for key, val in obj.items():
        val_type = type(val)
        if val_type in [int, float, str, bool]:
            lines.append(primitve(key, val))
        elif not val:
            lines.append(primitve(key, ""))
        elif val_type is list:
            for item in val:
                lines.extend(object(key, item))
        else:
            lines.extend(object(key, val))

    return lines


def object(key, val):
    lines = []

    lines.append("%s {" % key)
    lines.extend(object_props(val))
    lines.append("}")

    return lines


def primitve(key, val):
    return "%s = %s" % (key, value(val))
